File: hub/hub/schemas/tasks.py
"""Task schemas."""

import re
from datetime import datetime
from typing import Any, List, Optional

from pydantic import BaseModel, Field, field_validator, model_validator

from .common import RequestModel

# Matches generated ids of the form "{prefix}-{hex}", where the prefix is a short word (e.g.
# "task", "msg"). Deliberately does not pin the segment's width: `short_id()` widened from 8 hex
# characters to 12 on 2026-08-24 and ids at both widths are valid forever, because a segment is only
# ever generated and never parsed. Used to validate client-supplied ids so we only accept well-formed
# ones and reject anything that could be used for path traversal or to impersonate other entity
# types.
_TASK_ID_RE = re.compile(r"^[a-zA-Z][a-zA-Z0-9_-]{0,63}$")

_TASK_STATUSES = [
    "pending",
    "assigned",
    "in_progress",
    "blocked",
    "completed",
    "under_review",
    "revision_needed",
    "approved",
    "rejected",
]
_PRIORITIES = ["low", "medium", "high", "critical"]

# Restated from `hub.task_transitions.ENTRY_STATUSES`; the two are pinned together by
# `hub/tests/test_task_transitions.py`. Restated rather than imported because this module is a
# leaf that the schemas layer imports widely, and the transition module imports nothing.
_ENTRY_STATUSES = {"pending", "assigned"}
# Older writers name the assignee two other ways. Both models read them and neither
# declares them, so both must also *remove* them — see the validators below.
_ASSIGNEE_ALIASES = ("assigned_to", "assigned_agent")

# Restated from `hub.run_task_binding.POLICIES` for the same reason, and pinned to it by
# `hub/tests/test_run_task_binding.py`.
_DIVERGENCE_POLICIES = {"surface", "retry", "escalate"}


class TaskCreate(RequestModel):
    title: str = Field(max_length=256)
    description: str = Field(default="", max_length=10000)
    status: str = Field(default="pending", max_length=64)
    priority: str = Field(default="medium", max_length=64)
    assignee: Optional[str] = Field(default=None, max_length=64)
    assigner: Optional[str] = Field(default=None, max_length=64)
    requirements: Optional[List[Any]] = None
    # The requirements this task serves, by identifier. Checked, unlike `requirements`: an
    # identifier the project does not have is refused with the identifier named, rather than stored
    # as text that looks like a reference and resolves to nothing.
    requirement_ids: Optional[List[str]] = None
    # Which document's identifiers to resolve against. Identifiers are minted per document, so a
    # bare `FR-8` is ambiguous where two documents declare one; this is how a caller says which.
    spec_document: Optional[str] = Field(default=None, max_length=255)
    # Adds this task directly to a loop's queue (design D1, `2026-08-18-a-loop-writes-its-own-
    # queue`). Gated in `create_task_for_actor`: only the loop's own `AIJob.agent`, or the
    # operator, may supply this — never trusted from the field alone.
    loop_id: Optional[str] = Field(default=None, max_length=64)
    acceptance_criteria: Optional[List[Any]] = None
    deliverables: Optional[List[Any]] = None
    notes: Optional[Any] = None
    # Optional client-supplied id. When present, the Hub uses it instead of
    # generating one — this lets the MCP `create_task` tool return the same
    # id that the Hub stored, so subsequent get_task / update_task calls by
    # the agent find the task. Validated to the same shape as the CLI's
    # generate_id() output and the local Task model.
    id: Optional[str] = Field(default=None, max_length=64)

    @field_validator("id")
    @classmethod
    def _validate_id_shape(cls, v: Optional[str]) -> Optional[str]:
        if v is None:
            return None
        if not _TASK_ID_RE.match(v):
            raise ValueError(
                "id must start with a letter and contain only letters, "
                "digits, underscores, or hyphens (max 64 chars)"
            )
        return v

    @model_validator(mode="before")
    @classmethod
    def normalize_assignee_aliases(cls, data: Any) -> Any:
        if isinstance(data, dict):
            if data.get("assignee") is None:
                for key in _ASSIGNEE_ALIASES:
                    if data.get(key):
                        data = {**data, "assignee": data[key]}
                        break
            # Remove the alias keys whichever name won — a rolling upgrade emits the
            # canonical name *and* its alias in the same body, and stripping only when
            # `assignee` was absent handed the survivor to extra='forbid', refusing a
            # body the contract itself accepts.
            data = {k: v for k, v in data.items() if k not in _ASSIGNEE_ALIASES}
        return data

    @field_validator("status")
    @classmethod
    def validate_status(cls, v: str) -> str:
        # Creation is narrower than update: a task may only *start* at an entry point, or the
        # machine is walkable around by creating a task already `approved`
        # (openspec/changes/2026-08-10-task-transition-machine, design D10).
        if v not in _ENTRY_STATUSES:
            raise ValueError(f"a new task must start at one of {sorted(_ENTRY_STATUSES)}")
        return v

    @field_validator("priority")
    @classmethod
    def validate_priority(cls, v: str) -> str:
        if v not in _PRIORITIES:
            raise ValueError(f"priority must be one of {_PRIORITIES}")
        return v


class TaskUpdate(RequestModel):
    status: Optional[str] = Field(default=None, max_length=64)
    priority: Optional[str] = Field(default=None, max_length=64)
    # `None` means *clear it*, not *leave it alone* — the difference is carried by
    # `model_fields_set`, exactly as `escalation_agent` below carries it.
    # `update_task_for_actor` read this field as "unset when None" until F78, which made the
    # remedy `_guard_reviewer_is_not_the_author` names — "clear the assignee to review it
    # yourself" — unreachable, and unreachable *silently*: the PATCH returned 200 with the old
    # holder still in it.
    assignee: Optional[str] = Field(default=None, max_length=64)
    description: Optional[str] = Field(default=None, max_length=10000)
    notes: Optional[Any] = None
    # Requirements this task serves, added to whatever it already serves. Links are never removed
    # here: what work served a requirement is asked mostly about finished work, so the record has to
    # outlive the editing of the task.
    requirement_ids: Optional[List[str]] = None
    spec_document: Optional[str] = Field(default=None, max_length=255)
    divergence_policy: Optional[str] = Field(default=None, max_length=16)
    # Deliberately not `Optional[str] = None means leave alone` for this one: clearing an escalation
    # agent is a thing the operator must be able to do, and `""` is how they say it. Normalised to
    # NULL below so the column never holds an empty name.
    escalation_agent: Optional[str] = Field(default=None, max_length=64)
    # What the task is waiting for, when the operator parks it by hand. Ignored on any other status
    # change — leaving `blocked` always clears it, since a reason that outlives its block describes
    # something that already arrived.
    blocked_reason: Optional[str] = Field(default=None, max_length=2000)
    # `Task.loop_id` is write-once (design D14, `2026-08-18-a-loop-writes-its-own-queue`): a loop's
    # queue history has to be able to answer what work it was ever given, which reassignment would
    # break. Accepted here only so the service layer (`update_task_for_actor`) has a value to refuse
    # by name rather than by `extra="forbid"` silently swallowing it — see D14 for why this is
    # enforced in code and not a DB constraint (SQLite cannot drop one later).
    loop_id: Optional[str] = Field(default=None, max_length=64)

    @model_validator(mode="after")
    def blocking_by_hand_must_say_what_for(self) -> "TaskUpdate":
        """A hand-set block names what it is waiting for, or it is not accepted (R5).

        Required rather than optional because an unexplained block is the failure mode the status
        was introduced to fix. A card that says only "blocked" leaves the operator working out what
        they are holding up — which is exactly the position they were in when the task said
        `in_progress` and nothing was happening.

        A runtime block is not affected: it fills the reason from the question text, and does not
        come through this schema.
        """
        if self.status == "blocked" and not (self.blocked_reason or "").strip():
            raise ValueError("blocked_reason is required when setting a task to blocked")
        return self

    @field_validator("blocked_reason")
    @classmethod
    def normalise_blocked_reason(cls, v: Optional[str]) -> Optional[str]:
        return v.strip() or None if isinstance(v, str) else v

    @field_validator("divergence_policy")
    @classmethod
    def validate_divergence_policy(cls, v: Optional[str]) -> Optional[str]:
        if v is not None and v not in _DIVERGENCE_POLICIES:
            raise ValueError(f"divergence_policy must be one of {sorted(_DIVERGENCE_POLICIES)}")
        return v

    @field_validator("escalation_agent")
    @classmethod
    def normalise_escalation_agent(cls, v: Optional[str]) -> Optional[str]:
        return v.strip() or None if isinstance(v, str) else v

    @field_validator("assignee")
    @classmethod
    def normalise_assignee(cls, v: Optional[str]) -> Optional[str]:
        """`""` and `"  "` are the same claim as `null`: nobody holds this task.

        Without this the column grows a second falsy spelling of "unassigned", which every reader
        happens to survive today only because they all test Python truthiness — while the four
        `Task.assignee.isnot(None)` queries in the Hub would quietly start counting it as a holder.
        """
        return v.strip() or None if isinstance(v, str) else v

    @model_validator(mode="before")
    @classmethod
    def normalize_assignee_aliases(cls, data: Any) -> Any:
        if isinstance(data, dict):
            if data.get("assignee") is None:
                for key in _ASSIGNEE_ALIASES:
                    if data.get(key):
                        data = {**data, "assignee": data[key]}
                        break
            # This model never removed them at all, so *every* alias body it read was
            # then refused by extra='forbid' — the alias was accepted in principle and
            # rejected in fact. Same line as its sibling above.
            data = {k: v for k, v in data.items() if k not in _ASSIGNEE_ALIASES}
        return data

    @field_validator("status")
    @classmethod
    def validate_status(cls, v: Optional[str]) -> Optional[str]:
        if v is not None and v not in _TASK_STATUSES:
            raise ValueError(f"status must be one of {_TASK_STATUSES}")
        return v

    @field_validator("priority")
    @classmethod
    def validate_priority(cls, v: Optional[str]) -> Optional[str]:
        if v is not None and v not in _PRIORITIES:
            raise ValueError(f"priority must be one of {_PRIORITIES}")
        return v


class TaskDependencyRef(BaseModel):
    """A prerequisite or dependent named on `TaskResponse` — enough to render an edge without a
    second fetch: `id` to link to it, `title` and `status` to show what it is and whether it has
    cleared (`task-dependencies` design D3, task 7.1). `spec_document_id` is what lets a board
    draw a prerequisite outside its own document as an off-board reference naming that document
    (task 8.7) rather than a bare title with nowhere to point."""

    id: str = Field(max_length=64)
    title: str = Field(max_length=256)
    status: str = Field(max_length=32)
    spec_document_id: Optional[str] = Field(default=None, max_length=128)

    model_config = {"from_attributes": True}


class TaskIntegrationSummary(BaseModel):
    """What the most recent approval of this task did to the repository — merged, or skipped why.

    A trimmed echo of one row from `GET /tasks/{id}/integrations`, which existed already but told
    nobody unless they asked it directly. "Approving is what merges it" was true and silent: the
    approve response itself gave no sign whether a merge happened, was skipped (no main branch
    configured, a dirty checkout, nothing to merge), or is not applicable because the task never
    reached `approved`.
    """

    outcome: str = Field(max_length=16)
    reason: str = Field(default="", max_length=2000)
    commit_sha: Optional[str] = Field(default=None, max_length=64)
    target_branch: Optional[str] = Field(default=None, max_length=255)
    created_at: datetime

    model_config = {"from_attributes": True}


class TaskResponse(BaseModel):
    id: str = Field(max_length=128)
    project_id: str = Field(max_length=128)
    title: str = Field(max_length=256)
    description: str = Field(max_length=10000)
    status: str = Field(max_length=64)
    priority: str = Field(max_length=64)
    assignee: Optional[str] = Field(default=None, max_length=64)
    assigner: Optional[str] = Field(default=None, max_length=64)
    created_at: datetime
    updated: datetime
    # What the caller submitted, kept verbatim. Answers about traceability come from
    # `requirement_links`, never from here — this is the original, so a mis-parse is re-derivable.
    requirements: Optional[Any] = None
    # The requirements this task actually serves: `identifier`, `requirement_id`, `document_id`,
    # `state`, `anchor`, `key`, and — read from the document rather than stored — `statement` and
    # `modal`. `statement` is null where the document no longer words the requirement, which is
    # what a retired one is. `has_rejected_evidence`, `rejected_evidence_count` and
    # `latest_rejection_reason` cover a gap `state` cannot: evidence rejected at the requirement's
    # current digest reads identically to a requirement nobody has attempted, because
    # `requirement_coverage` has no precedence level for "tried and rejected".
    requirement_links: List[Any] = Field(default_factory=list)
    # The read side of `TaskCreate`/`TaskUpdate.requirement_ids`, which were accepted and reported
    # nowhere — so a caller could not confirm what was recorded, and anyone diagnosing why work did
    # not merge saw a task that appeared tied to nothing while the links governing the merge
    # existed. Identifiers, in the form they are submitted in, so what is read back can be sent
    # again; derived from `requirement_links` and never stored.
    requirement_ids: List[str] = Field(default_factory=list)
    # References that named no requirement this project has, with their original text. Visible
    # rather than dropped, because a task that quietly lost a reference it used to have is the
    # failure the migration exists to prevent.
    unresolved_requirements: List[Any] = Field(default_factory=list)
    # `contract`-rigor requirements this task serves that were not verified at the moment this
    # response answers an approval — identifier, state, remedy, the same shape a `gate` refusal
    # names its blockers with. Empty on every response but the one that just approved a task under a
    # document that reports rather than refuses; never persisted, because it describes this moment,
    # not a record (`2026-08-13-a-gate-that-only-evidence-opens`, task 5.5).
    approval_report: List[Any] = Field(default_factory=list)
    acceptance_criteria: Optional[Any] = None
    deliverables: Optional[Any] = None
    notes: Optional[Any] = None
    assignee_status: Optional[str] = Field(default=None, max_length=64)
    assignee_status_msg: Optional[str] = Field(default=None, max_length=10000)
    assignee_last_seen: Optional[datetime] = None
    divergence_policy: str = Field(default="surface", max_length=16)
    escalation_agent: Optional[str] = Field(default=None, max_length=64)
    # Whether a run bound to this task ended without moving it and nothing has since. Computed,
    # not stored: the durable record is the divergence row, and a second copy on the task would be
    # one more thing that can disagree with it.
    has_open_divergence: bool = False
    # What a `blocked` task is waiting for. NULL on every other status. Since a blocked task stays
    # in the in_progress column rather than moving to one of its own (R3), this is most of what
    # tells the operator the card is waiting on them.
    blocked_reason: Optional[str] = Field(default=None, max_length=2000)
    # What a run bound to this task is waiting on the operator to answer, right now — in the same
    # words `blocked_reason` uses on a task that parked.
    #
    # The ordinary case is now the status: the agent-facing question routes park the asking run's
    # task at ask time (F14). This covers the two waits the status cannot — a task that could not
    # park (`under_review`, `pending`, `assigned`), and a batch whose first answer released the task
    # while the run waits on the rest. Computed per request, never stored — the question row is the
    # record.
    awaiting_answer_reason: Optional[str] = Field(default=None, max_length=2000)
    # Which document this work is against, and — for a task the document itself declared — which of
    # its declared units this is. Written since migration `0071` and exposed nowhere, so nothing
    # above the database layer could tell a declared task from a hand-made one, or get from a task
    # to the specification it implements.
    spec_document_id: Optional[str] = Field(default=None, max_length=64)
    spec_task_key: Optional[str] = Field(default=None, max_length=128)
    # Which loop's queue this task belongs to. `TaskCreate` has accepted `loop_id` since the column
    # existed and the response never carried it back, so `POST /tasks {"loop_id": …}` answered 201
    # with `loop_id: null` while the loop's own summary already counted the task in its queue. The
    # write worked; only the reply denied it, and a caller had no way to confirm from the create
    # call that the task had joined the loop.
    loop_id: Optional[str] = Field(default=None, max_length=64)
    # The most recent attempt to integrate this task's approved work, or null where none has ever
    # been made (every non-`approved` task, and one whose approval predates this field existing on
    # older rows the migration never touched). The full history stays at
    # `GET /tasks/{id}/integrations`; this is only ever the newest row of it.
    latest_integration: Optional[TaskIntegrationSummary] = None
    # `TaskDependency` read from both ends (`task-dependencies` design D3, task 7.1). Never stored
    # here — derived per request from the join table, the same way `requirement_links` is derived
    # from `task_requirement_links` rather than kept as a second copy.
    prerequisites: List[TaskDependencyRef] = Field(default_factory=list)
    dependents: List[TaskDependencyRef] = Field(default_factory=list)
    # One of `"gated"` (a prerequisite is not yet approved), `"gated_on_rejected"` (a prerequisite
    # was rejected and will not clear on its own), `"running_on_regressed"` (this task is already
    # `in_progress` but a prerequisite no longer reads `approved` — flagged, not stopped, per design
    # D8), or `None` when nothing about this task's dependencies is worth surfacing. Derived per
    # request, not stored (task 7.2) — a stored readiness column is a denormalised join that goes
    # stale the moment a prerequisite's status changes under it (design D1).
    dependency_state: Optional[str] = Field(default=None, max_length=32)

    model_config = {"from_attributes": True}

File: hub/ui/src/api/tasks.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import { getJson, patchJson, postJson } from './client'
import { useConfigStore } from '@/store/configStore'

export interface Task {
  id: string
  project_id: string
  title: string
  description?: string
  status: string
  priority: string
  assignee?: string
  assigner?: string
  assignee_status?: string | null
  assignee_status_msg?: string | null
  assignee_last_seen?: string | null
  requirements?: string[]
  acceptance_criteria?: string[]
  deliverables?: string[]
  notes?: string
  created_at: string
  updated: string
  /** What happens when a run bound to this task ends without the task moving. */
  divergence_policy: DivergencePolicy
  escalation_agent?: string | null
  /** A run dropped this task and nothing has moved it since. */
  has_open_divergence: boolean
  /**
   * What a `blocked` task is waiting for, in words. Null on every other status.
   *
   * A blocked task stays in the In Progress column rather than moving to one of its own, so this
   * is most of what tells the operator the card is waiting on *them* rather than merely stalled.
   */
  blocked_reason?: string | null
  /**
   * What a run bound to this task is waiting on *you* to answer, right now — in the same words
   * `blocked_reason` uses on a task that parked.
   *
   * The ordinary wait now moves the status: the task parks the moment the agent asks (F14). This
   * covers the two waits the status cannot — a task that could not park (`under_review`, `pending`,
   * `assigned`), and a batch whose first answer released the task while the run waits on the rest.
   * Derived per request, never stored.
   */
  awaiting_answer_reason?: string | null
  /** The specification document this work is against, and — where a document declared this task —
   *  the key it was declared under. */
  spec_document_id?: string | null
  spec_task_key?: string | null
  /**
   * The requirement identifiers this task serves, in the form they are submitted in.
   *
   * `requirements` is the caller's verbatim prose and can say things no identifier can;
   * these are the checked links the approval gate actually enforces. The board showed only
   * the prose, so a card could not tell you whether it was tied to the specification at all.
   */
  requirement_ids?: string[]
  requirement_links?: RequirementLink[]
  unresolved_requirements?: { reference: string; reason: string }[]
  /** `TaskDependency` read from both ends — enough to render an edge without a second fetch
   *  (`task-dependencies` task 7.1). Absent on a response that never attached dependencies. */
  prerequisites?: TaskDependencyRef[]
  dependents?: TaskDependencyRef[]
  /** `"gated"` | `"gated_on_rejected"` | `"running_on_regressed"` | `null` — derived per request,
   *  never stored (task 7.2, design D1). */
  dependency_state?: string | null
}

/** A prerequisite or dependent named on a `Task` — enough to draw an edge without fetching the
 *  other end of it. `spec_document_id` is null for a hand-made task and is what lets a board draw
 *  a prerequisite outside its own document as an off-board reference naming that document
 *  (`task-dependencies` task 8.7), rather than a bare title with nowhere to point. */
export interface TaskDependencyRef {
  id: string
  title: string
  status: string
  spec_document_id?: string | null
}

/** One checked tie between a task and a requirement. `statement` is null where the document no
 *  longer words the requirement, which is what a retired one is. */
export interface RequirementLink {
  identifier: string
  requirement_id: string
  document_id: string
  state: string
  anchor?: string | null
  key?: string | null
  statement?: string | null
  modal?: string | null
  /** True only for evidence rejected against the requirement's *current* digest. Kept independent
   *  of coverage's `rejected` state: this stays true even after a later acceptance moves coverage
   *  on to `verified` — see `hub/hub/api/v1/tasks.py`. */
  has_rejected_evidence?: boolean
  rejected_evidence_count?: number
  latest_rejection_reason?: string | null
}

export type DivergencePolicy = 'surface' | 'retry' | 'escalate'

export const DIVERGENCE_POLICY_LABELS: Record<DivergencePolicy, string> = {
  surface: 'Tell me',
  retry: 'Try again once',
  escalate: 'Hand to another agent',
}

/**
 * What approving a task did to the repository — including the times it did nothing.
 *
 * A skipped merge with a stated reason is the answer to "my approved work is not on main", so the
 * skips matter as much as the merges and are not filtered out here.
 */
export interface TaskIntegration {
  id: string
  commit_sha: string | null
  source_branch: string | null
  target_branch: string | null
  outcome: 'merged' | 'skipped' | 'failed'
  reason: string
  /** Commits that landed alongside `commit_sha` because a merge brings in a commit's whole
   *  ancestry, not its diff alone (F58). Empty means nothing rode along. */
  rode_along_commits: string[]
  mechanism: string
  actor_kind: string
  actor: string
  created_at: string
}

export function useTaskIntegrations(taskId: string, enabled: boolean) {
  const { isConfigured, selectedProjectId: projectId } = useConfigStore()
  return useQuery<{ integrations: TaskIntegration[] }>({
    queryKey: ['project', projectId, 'task', taskId, 'integrations'],
    queryFn: () =>
      getJson<{ integrations: TaskIntegration[] }>(
        `/api/v1/projects/${projectId}/tasks/${taskId}/integrations`,
      ),
    enabled: isConfigured && !!projectId && enabled,
  })
}

/**
 * What approving this task *would* write, before it is approved (F9).
 *
 * Approval is the only act in the product that changes the operator's own repository — it
 * cherry-picks the commit named by each accepted piece of evidence into the project's main branch.
 * The refusal path already explained itself ("no accepted evidence names a commit"); the
 * *successful* path announced nothing at all, and an operator clicking approve on a task board was
 * writing to their main branch without being told.
 *
 * Read-only and cheap on the server: same source as the merge (`integration_targets` plus
 * `Project.main_branch`), no git subprocess, no conflict probe. This is a sentence, not a gate.
 */
export interface TaskIntegrationPreview {
  task_id: string
  main_branch: string | null
  targets: { commit_sha: string; source_branch: string | null }[]
  will_merge: boolean
  /** Why nothing will be merged. Empty when something will. */
  reason: string
}

export function useTaskIntegrationPreview(taskId: string, enabled: boolean) {
  const { isConfigured, selectedProjectId: projectId } = useConfigStore()
  return useQuery<TaskIntegrationPreview>({
    queryKey: ['project', projectId, 'task', taskId, 'integration-preview'],
    queryFn: () =>
      getJson<TaskIntegrationPreview>(
        `/api/v1/projects/${projectId}/tasks/${taskId}/integration-preview`,
      ),
    enabled: isConfigured && !!projectId && enabled,
  })
}

/**
 * Attempt the merge again for an approved task whose work is not in the product.
 *
 * Approving again cannot re-run it — restating a status is a no-op — so without this the reason
 * text on the note asks for a remediation that accomplishes nothing.
 */
export function useRetryTaskIntegration(taskId: string) {
  const queryClient = useQueryClient()
  const { selectedProjectId: projectId } = useConfigStore()
  return useMutation({
    mutationFn: () =>
      postJson<{ integrations: TaskIntegration[] }>(
        `/api/v1/projects/${projectId}/tasks/${taskId}/integrations/retry`,
      ),
    onSuccess: () => {
      queryClient.invalidateQueries({
        queryKey: ['project', projectId, 'task', taskId, 'integrations'],
      })
    },
  })
}

/** `loopId` and `excludeArchivedCompleted` both scope the same query on the Hub (an `elif` chain,
 *  `hub/hub/api/v1/tasks.py`), so passing both is not meaningful — callers pick one. */
export function useTasks(options?: { excludeArchivedCompleted?: boolean; loopId?: string }) {
  const { isConfigured, selectedProjectId: projectId } = useConfigStore()
  const excludeArchivedCompleted = options?.excludeArchivedCompleted ?? false
  const loopId = options?.loopId
  return useQuery<Task[]>({
    queryKey: ['project', projectId, 'tasks', { excludeArchivedCompleted, loopId }],
    queryFn: () => {
      const params = new URLSearchParams()
      if (loopId) params.set('loop_id', loopId)
      else if (excludeArchivedCompleted) params.set('exclude_archived_completed', 'true')
      const qs = params.toString()
      return getJson<Task[]>(`/api/v1/projects/${projectId}/tasks${qs ? `?${qs}` : ''}`)
    },
    enabled: isConfigured && !!projectId,
  })
}

/** Every task one specification document declared, regardless of status or the document's own
 *  phase — an explicit scope never hides anything. Its own hook rather than `useTasks({
 *  specDocumentId })`: the two exist for different callers (one small link fetching one document's
 *  tasks; the whole board fetching everything it will render). */
export function useDocumentTasks(documentId: string | null) {
  const { isConfigured, selectedProjectId: projectId } = useConfigStore()
  return useQuery<Task[]>({
    queryKey: ['project', projectId, 'tasks', { spec_document_id: documentId }],
    queryFn: () =>
      getJson<Task[]>(
        `/api/v1/projects/${projectId}/tasks?spec_document_id=${encodeURIComponent(documentId ?? '')}`,
      ),
    enabled: isConfigured && !!projectId && !!documentId,
  })
}

/** One edge in a board's graph: `task_id` depends on `depends_on_task_id`. */
export interface TaskBoardEdge {
  task_id: string
  depends_on_task_id: string
}

export interface TaskBoard {
  spec_document_id: string | null
  tasks: Task[]
  edges: TaskBoardEdge[]
}

/**
 * One document's tasks and the edges between them, in one call (`task-dependencies` task 7.3,
 * design D9). `specDocumentId` of `null` is not "nothing selected" — it names the standing
 * "no document" board (every hand-made task, which per design D5 can never have an edge), so this
 * stays enabled for it rather than treating null as disabled the way `useDocumentTasks` does.
 */
export function useTaskBoard(specDocumentId: string | null) {
  const { isConfigured, selectedProjectId: projectId } = useConfigStore()
  return useQuery<TaskBoard>({
    queryKey: ['project', projectId, 'tasks', 'board', specDocumentId],
    queryFn: () => {
      const qs = specDocumentId ? `?spec_document_id=${encodeURIComponent(specDocumentId)}` : ''
      return getJson<TaskBoard>(`/api/v1/projects/${projectId}/tasks/board${qs}`)
    },
    enabled: isConfigured && !!projectId,
  })
}

/** The picker: every board that has tasks, with outstanding counts (task 7.4, design D9). */
export interface TaskBoardSummary {
  spec_document_id: string | null
  title: string | null
  total: number
  outstanding: number
}

export function useTaskBoards() {
  const { isConfigured, selectedProjectId: projectId } = useConfigStore()
  return useQuery<{ boards: TaskBoardSummary[] }>({
    queryKey: ['project', projectId, 'tasks', 'boards'],
    queryFn: () => getJson<{ boards: TaskBoardSummary[] }>(`/api/v1/projects/${projectId}/tasks/boards`),
    enabled: isConfigured && !!projectId,
  })
}

export function useUpdateTask() {
  const queryClient = useQueryClient()
  const { selectedProjectId: projectId } = useConfigStore()
  return useMutation({
    // `blocked_reason` is required by the Hub when the status is `blocked`, and ignored otherwise —
    // so it is sent only when present rather than as a null the validator would still have to
    // reject.
    mutationFn: ({
      id,
      status,
      blocked_reason,
    }: {
      id: string
      status: string
      blocked_reason?: string
    }) =>
      patchJson<Task>(`/api/v1/projects/${projectId}/tasks/${id}`, {
        status,
        ...(blocked_reason ? { blocked_reason } : {}),
      }),
    onSuccess: () => queryClient.invalidateQueries({ queryKey: ['project', projectId, 'tasks'] }),
  })
}

/**
 * How this task's neglect should be answered.
 *
 * Separate from `useUpdateTask` because it is a different kind of act: that one moves the task
 * through its lifecycle and can be refused by the transition machine, this one records a standing
 * instruction and cannot. Sharing a mutation would put both under one pending state, so setting a
 * policy would look like a status change in flight.
 */
export function useSetDivergenceHandling() {
  const queryClient = useQueryClient()
  const { selectedProjectId: projectId } = useConfigStore()
  return useMutation({
    mutationFn: ({
      id,
      ...fields
    }: {
      id: string
      divergence_policy?: DivergencePolicy
      escalation_agent?: string | null
    }) => patchJson<Task>(`/api/v1/projects/${projectId}/tasks/${id}`, fields),
    onSuccess: () => queryClient.invalidateQueries({ queryKey: ['project', projectId, 'tasks'] }),
  })
}

export interface RunDivergence {
  id: string
  run_id: string
  agent: string
  task_id: string
  task_status_at_end: string
  run_exit_status: string
  policy_applied: DivergencePolicy
  outcome: 'surfaced' | 'retried' | 'escalated'
  response_run_id?: string | null
  previous_assignee?: string | null
  created_at: string
  resolved_at?: string | null
}

/** Runs that ended holding work nobody moved — including ones since picked up. */
export function useDivergences(openOnly = false) {
  const { isConfigured, selectedProjectId: projectId } = useConfigStore()
  return useQuery<RunDivergence[]>({
    queryKey: ['project', projectId, 'divergences', openOnly],
    queryFn: () =>
      getJson<RunDivergence[]>(
        `/api/v1/projects/${projectId}/tasks/divergences/recent?open_only=${openOnly}`,
      ),
    enabled: isConfigured && !!projectId,
  })
}

/**
 * Start a run bound to a task.
 *
 * The binding is what makes the run answerable at its boundary, and it is the Hub that sets it —
 * so this is a trigger carrying `task_id`, not a task update. The task moves to in_progress by
 * itself, without the agent being asked.
 */
export function useStartWorkOnTask() {
  const queryClient = useQueryClient()
  const { selectedProjectId: projectId } = useConfigStore()
  return useMutation({
    mutationFn: ({ taskId, agent, title }: { taskId: string; agent: string; title: string }) =>
      postJson<{ run_id?: string; conversation_id: string }>(
        `/api/v1/projects/${projectId}/agent/trigger`,
        {
          agent,
          message: `Work on task ${taskId}: ${title}`,
          task_id: taskId,
        },
      ),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['project', projectId, 'tasks'] })
      queryClient.invalidateQueries({ queryKey: ['project', projectId, 'agents'] })
    },
  })
}

/** `{ from_status: [reachable...] }` for the operator, as the Hub declares it. */
export type AllowedTransitions = Record<string, string[]>

/**
 * The transition map, fetched once for the operator rather than per task.
 *
 * The legal set depends on who is asking, so it deliberately does not ride on the task response —
 * a resource that varies by asker breaks what every cache assumes about it, this query key
 * included. And it is not per-task: forty cards in the same status have one answer, not forty.
 * See design D13 of `openspec/changes/2026-08-10-task-transition-machine`.
 *
 * Serving it from the same declaration the Hub enforces is the point. A copy of the map here would
 * drift, and the first symptom would be the card offering a move that is then refused.
 */
export function useAllowedTransitions() {
  const { isConfigured, selectedProjectId: projectId } = useConfigStore()
  return useQuery<{ actor_kind: string; transitions: AllowedTransitions }>({
    queryKey: ['project', projectId, 'task-transitions'],
    queryFn: () =>
      getJson<{ actor_kind: string; transitions: AllowedTransitions }>(
        `/api/v1/projects/${projectId}/tasks/transitions/allowed`,
      ),
    enabled: isConfigured && !!projectId,
    // The map changes only when the Hub is redeployed, so refetching it per window focus is noise.
    staleTime: Infinity,
  })
}
